src/Controller/ModulesController.php

```diff
@@ -523,4 +523,117 @@ public function getSchemaForIndex($objectType): array
 
         return $schema;
     }
+
+    /**
+     * List categories for the object type.
+     *
+     * @return \Cake\Http\Response|null
+     */
+    public function listCategories()
+    {
+        $this->viewBuilder()->setTemplate('categories');
+
+        $this->request->allowMethod(['get']);
+        $query = $this->request->getQueryParams() + [
+            'filter' => [
+                'type' => $this->objectType,
+            ],
+            'page_size' => 500,
+        ];
+
+        try {
+            $response = $this->apiClient->get('/model/categories', $query);
+        } catch (BEditaClientException $e) {
+            $this->log($e, 'error');
+            $this->Flash->error($e->getMessage(), ['params' => $e]);
+
+            return $this->redirect(['_name' => 'dashboard']);
+        }
+
+        $resources = [];
+        foreach ((array)$response['data'] as $category) {
+            $resources[$category['id']] = $category;
+        }
+
+        $grouped = [
+            '_' => [],
+        ];
+        foreach ($resources as $category) {
+            if (empty($category['attributes']['parent_id'])) {
+                $grouped['_'][] = $category['id'];
+            } else {
+                $grouped[$category['attributes']['parent_id']][] = $category['id'];
+            }
+        }
+
+        $this->set(compact('resources', 'grouped'));
+        $this->set('meta', (array)$response['meta']);
+        $this->set('links', (array)$response['links']);
+        $this->set('schema', $this->Schema->getSchema());
+        $this->set('properties', $this->Properties->indexList('categories'));
+        $this->set('filter', $this->Properties->filterList('categories'));
+        $this->set('object_types', [$this->objectType]);
+
+        return null;
+    }
```

Comment:
the logic in this function is very similar to `index` (except for `$query`). you could put it in a function inside a component or in a model

Reply:
Yes, I was not sure how to abstract this part. The component solution should fit I think